execution-service/src/main/java/org/ihtsdo/rvf/execution/service/MysqlFailuresExtractor.java

```diff
@@ -0,0 +1,187 @@
+
+@Service
+public class MysqlFailuresExtractor {
+
+    private final Logger logger = LoggerFactory.getLogger(MysqlFailuresExtractor.class);
+
+    @Resource(name = "dataSource")
+    private BasicDataSource dataSource;
+
+    @Value("${rvf.qa.result.table.name}")
+    private String qaResultTableName;
+
+    @Autowired
+    private WhitelistService whitelistService;
+
+    public void extractTestResults(final List<TestRunItem> items, final MysqlExecutionConfig config, List<Assertion> assertions) throws SQLException {
+        Map<UUID, Long> uuidToAssertionIdMap = assertions.stream().collect(Collectors.toMap(Assertion::getUuid, assertion -> assertion.getAssertionId()));
+        try (Connection connection = dataSource.getConnection()) {
+            Map<String, Integer> assertionIdToTotalFailureMap = getAssertionIdToTotalFailureMap(connection, config);
+            if (whitelistService.isWhitelistDisabled()) {
+                extractTestResults(connection, items, config, uuidToAssertionIdMap, assertionIdToTotalFailureMap);
+            } else {
+                validateFailuresAndExtractTestResults(connection, items, config, uuidToAssertionIdMap, assertionIdToTotalFailureMap);
+            }
+        }
+    }
+
+    private void extractTestResults(Connection connection, List<TestRunItem> items, MysqlExecutionConfig config, Map<UUID, Long> uuidToAssertionIdMap, Map<String, Integer> assertionIdToTotalFailureMap) throws SQLException {
+        for (TestRunItem item : items) {
+            String key = String.valueOf(uuidToAssertionIdMap.get(item.getAssertionUuid()));
+            if (assertionIdToTotalFailureMap.containsKey(key)) {
+                item.setFailureCount(Long.valueOf(assertionIdToTotalFailureMap.get(key)));
+                item.setFirstNInstances(fetchFailureDetails(connection, config.getExecutionId(), uuidToAssertionIdMap.get(item.getAssertionUuid()), config.getFailureExportMax(), null, null));
+            } else {
+                item.setFailureCount(0L);
+                item.setFirstNInstances(null);
+            }
+        }
+    }
+
+    private void validateFailuresAndExtractTestResults(Connection connection, List<TestRunItem> items, MysqlExecutionConfig config, Map<UUID, Long> uuidToAssertionIdMap, Map<String, Integer> assertionIdToTotalFailureMap) throws SQLException {
+        for (TestRunItem item : items) {
+            String key = String.valueOf(uuidToAssertionIdMap.get(item.getAssertionUuid()));
+            if (assertionIdToTotalFailureMap.containsKey(key)) {
+                int batch_counter = 0;
+                int total_failures_extracted = 0;
+                int total_whitelistedItem_extracted = 0;
+                int total_failures = assertionIdToTotalFailureMap.get(key);
+                List<FailureDetail> firstNInstances = new ArrayList<>();
+                while(batch_counter * config.getFailureExportMax() < total_failures && total_failures_extracted < config.getFailureExportMax()) {
```

Comment:
Very nice implementation. Loading batches of failures and checking whitelist until we have the first N non-whitelisted failures extracted 👍